### PR TITLE
Pardot OAuth Integration

### PR DESCRIPTION
**Pull request by tim-dot-org**
Pardot changed their authentication flow from using an api key obtained through username/password to OAuth, which has multiple authentication flows. We're using the JWT (json web token) flow because it allows non-interactive logins. Setup on the pardot/salesforce config side took a long time, but now our integration allows us to use the new authentication. I added one private key (pardot_private_key) to the secrets manager in development and production environments (this script should only be running in the production daemon environment). The certificate (private key) expires in 100 years, so we'll have to update this then.

## Links

- documentation [jwt flow](https://help.salesforce.com/articleView?id=sf.remoteaccess_oauth_jwt_flow.htm)
- helpful [link](https://thespotforpardot.com/2021/02/02/pardot-api-and-getting-ready-with-salesforce-sso-users-part-3b-connecting-to-pardot-api-from-code/)
- jira ticket: [PLC-1240](https://codedotorg.atlassian.net/browse/PLC-1240)

## Testing story
Tested manually that we can access pardot as before:
```
[development] dashboard > require "cdo/contact_rollups/v2/pardot.rb"
GetSecretValue: development/cdo/pardot_private_key
=> true
[development] dashboard > PardotV2.retrieve_pardot_ids_by_email('tim@code.org')
=> ["88970755"]
```

### CODE REVIEW

File: lib/cdo/contact_rollups/v2/pardot_helpers.rb [hunk 1]
```diff
@@ -38,33 +40,51 @@ def log(s)
...
+    # build token payload
+    payload = {
+      # connected app client id
+      "iss": "3MVG9fMtCkV6eLhej.9tKBIE6OLmMCsxJAqIfy_eeSC1UaUR4rL0jkzUQOSRAhzfpHmUxUcuBp2JabX1ZOl2p",
+      # always login.salesforce.com
+      "aud": "https://login.salesforce.com",
+      # pardot account email
+      "sub": "plc-emails@code.org",
+      # less than 3 minutes after now
+      "exp": (Time.now + 2.minutes).to_i
```
Comment (bencodeorg) [inline lib/cdo/contact_rollups/v2/pardot_helpers.rb:63]: Probably a basic question, but would we just get new tokens repeatedly during a long-running job to send updates to Pardot? Is this similar to what happened previously?

**Review by bencodeorg — APPROVED** [review]

File: lib/cdo/contact_rollups/v2/pardot_helpers.rb (hunk 1 above)
Comment (tim-dot-org) [reply lib/cdo/contact_rollups/v2/pardot_helpers.rb:63]: yep, each individual request to pardot will regenerate the token if it fails to auth, which is how it was set up before

Comment (bencodeorg) [reply lib/cdo/contact_rollups/v2/pardot_helpers.rb:63]: Ok cool, am I reading the old version correctly that we typically reauthorized after an hour? Just better practice to do it more frequently? Just asking for my understanding.

Comment (tim-dot-org) [reply lib/cdo/contact_rollups/v2/pardot_helpers.rb:63]: oh yeah, its a requirement on salesforce's end for the exp field to be less than 3 min https://help.salesforce.com/s/articleView?id=sf.remoteaccess_oauth_jwt_flow.htm&type=5

**Pull request merged**